pkg/messaging/mqtt/pubsub_test.go

```diff
@@ -0,0 +1,426 @@
+	for _, tc := range cases {
+		expectedMsg := messaging.Message{
+			Publisher: "clientID11",
+			Channel:   tc.channel,
+			Subtopic:  tc.subtopic,
+			Payload:   tc.payload,
+		}
+		err := pubsub.Publish(topic, expectedMsg)
+		assert.Nil(t, err, fmt.Sprintf("%s: got unexpected error: %s\n", tc.desc, err))
+
+		receivedMsg := <-msgChan
+		if tc.payload == nil {
+			assert.Equal(t, 0, len(receivedMsg.Payload), fmt.Sprintf("%s: expected %+v got %+v\n", tc.desc, tc.payload, receivedMsg.Payload))
+		} else {
+			assert.Equal(t, tc.payload, receivedMsg.Payload, fmt.Sprintf("%s: expected %+v got %+v\n", tc.desc, tc.payload, receivedMsg.Payload))
+		}
+	}
+}
+
+func TestSubscribe(t *testing.T) {
+	msgChan := make(chan messaging.Message)
+
+	// Creating client to Publish messages to subscribed topic.
+	client, err := newClient(address, "mainflux", 30*time.Second)
+	assert.Nil(t, err, fmt.Sprintf("got unexpected error: %s", err))
+
+	t.Cleanup(func() {
+		client.Unsubscribe()
+		client.Disconnect(5)
+	})
+
+	cases := []struct {
+		desc     string
+		topic    string
+		clientID string
+		err      error
+		handler  messaging.MessageHandler
+	}{
+		{
+			desc:     "Subscribe to a topic with an ID",
+			topic:    topic,
+			clientID: "clientid1",
+			err:      nil,
+			handler:  handler{false, "clientid1", &msgChan},
+		},
+		{
+			desc:     "Subscribe to the same topic with a different ID",
+			topic:    topic,
+			clientID: "clientid2",
+			err:      nil,
+			handler:  handler{false, "clientid2", &msgChan},
+		},
+		{
+			desc:     "Subscribe to an already subscribed topic with an ID",
+			topic:    topic,
+			clientID: "clientid1",
+			err:      nil,
+			handler:  handler{false, "clientid1", &msgChan},
+		},
+		{
+			desc:     "Subscribe to a topic with a subtopic with an ID",
+			topic:    fmt.Sprintf("%s.%s", topic, subtopic),
+			clientID: "clientid1",
+			err:      nil,
+			handler:  handler{false, "clientid1", &msgChan},
+		},
+		{
+			desc:     "Subscribe to an already subscribed topic with a subtopic with an ID",
+			topic:    fmt.Sprintf("%s.%s", topic, subtopic),
+			clientID: "clientid1",
+			err:      nil,
+			handler:  handler{false, "clientid1", &msgChan},
+		},
+		{
+			desc:     "Subscribe to an empty topic with an ID",
+			topic:    "",
+			clientID: "clientid1",
+			err:      mqtt_pubsub.ErrEmptyTopic,
+			handler:  handler{false, "clientid1", &msgChan},
+		},
+		{
+			desc:     "Subscribe to a topic with empty id",
+			topic:    topic,
+			clientID: "",
+			err:      mqtt_pubsub.ErrEmptyID,
+			handler:  handler{false, "", &msgChan},
+		},
+	}
+	for _, tc := range cases {
+		err = pubsub.Subscribe(tc.clientID, tc.topic, tc.handler)
+		assert.Equal(t, err, tc.err, fmt.Sprintf("%s: expected: %s, but got: %s", tc.desc, err, tc.err))
+
+		if tc.err == nil {
+			token := client.Subscribe(tc.topic, qos, mqttHandler(handler{false, "clientID1", &msgChan}))
+			assert.Nil(t, token.Error(), fmt.Sprintf("got unexpected error: %s", token.Error()))
```

Comment:
Why do we still have this - what's the purpose of it in this test and why does the test get stuck if we remove it (hint: because handling in `pubsub.Subscribe` fails, but you need to figure out why)?

Reply:
Because when we create a new client in the first line of the function, we are not setting any handler. And since we're no longer using default message handler, so we need the client.Subscribe() function to set the handler for the client.
One other function to set a message handler for the mqtt-client, without subscribing is client.AddRoute(). I used that but that also resulted in not receiving message on channel.
Then, I also added the client.Connect(), but that is returning error `not connected`.
That's why I stuck with the client.Subscribe() technique😅